3scale/porta
# THREESCALE-9492: Bot protection: Remove suspicious only mode

**What this PR does / why we need it**:

This comes from https://github.com/3scale/porta/pull/3620. That PR was closed because it's not worth to reinvent the wheel and maintain our own spam protection system. So let's give it another try. In this PR the old Spam protection is removed and replaced by a much simpler bot protection based on reCaptcha v3.

There's no **Suspicious only** mode anymore. Now the user can only choose between **None** and **reCAPTCHA**:

![image](https://github.com/3scale/porta/assets/17052241/b75bb784-bb21-4658-bee1-1712ecbcdd1c)

All tenants configured to **Suspicious only** are now treated as **reCAPTCHA**.

ReCaptcha v3 doesn't add any prompt or expects the user to perform any action. The script gathers some information and sends it to reCaptcha in order to compute a value between 0 and 1 on how likely is the client human. A value of 0 means reCaptcha is sure the client is a bot, while 1 means the client is human.

When the reCaptcha challenge fails, the received data is rejected and a flash error is returned. This PR also fixes the UI, since the deployed version isn't returning the proper flash message in all screens.

**Which issue(s) this PR fixes** 

[THREESCALE-9492](https://issues.redhat.com/browse/THREESCALE-9492)

**Verification steps** 

As far as I know, there's no way to force reCaptcha to return a fake result when working on a development environment, so it basically always give me a high score and will probably be the same for reviewers. The only way to make it fail is to set `recaptcha_min_bot_score` ([here](https://github.com/3scale/porta/blob/318aa71899ad4b9d24dd288ce5ef5a95bd3eb1ca/config/examples/settings.yml#L26)) to `0.99` temporary. That way even high scores are considered bots and you can actually see the error message in the UI.

**Design**

The old spam protection code is removed, and replaced by a much more simpler version of it:

- `ThreeScale::BotProtection::LEVELS` ([code](https://github.com/3scale/porta/blob/6a83d7c5085cd9ea5c7be8d94373dbaf8bbe6995/app/lib/three_scale/bot_protection.rb#L5)): Available protection levels. Currently, only `:none` and `:captcha.`
- `ThreeScale::BotProtection::Controller` ([code](https://github.com/3scale/porta/blob/6a83d7c5085cd9ea5c7be8d94373dbaf8bbe6995/app/lib/three_scale/bot_protection/controller.rb)): A module to be included from controllers. It implements the method `bot_check` which returns `true` when the client is human and `false` when the client is a bot.
  - This module is included from:
    - `DeveloperPortal::SignupController`
    - `DeveloperPortal::Admin::Account::PasswordsController`
    - `Provider::SignupsController`
- `ThreeScale::BotProtection::Form` ([code](https://github.com/3scale/porta/blob/910b283e3115dcb6d351029d7cffc1c8e4fdff96/app/lib/three_scale/bot_protection/form.rb)): A module to be included from form builders. It implements the method `bot_protection_inputs` which returns the reCaptcha inputs when bot protection is enabled, or nothing otherwise.
  - This module is included from:
    - `ThreeScale::SemanticFormBuilder`
    - `Liquid::Forms::BotProtected`

**How to review**

Since reviewing 60 files is a nightmare, I created a few commits to split changes into smaller and easier to review change sets:

- Remove old logic (https://github.com/3scale/porta/pull/3645/commits/f9a7c4d2c9d1428a24e700f88c0a9d279b437be9): This one is pretty straightforward, just removing old code.
- Update recaptcha gem (https://github.com/3scale/porta/pull/3645/commits/7b216fa437865dd55737f544c519ed2277683c81): Easy and self-explanatory one.
- Add new logic (https://github.com/3scale/porta/pull/3645/commits/ce22cfa82db8b55697b1c6f52eabffc9c42b98d5): This one adds the new logic explained above
- UI updates (https://github.com/3scale/porta/pull/3645/commits/6f0b42e9272591afdd0770ecb66cea7a42abf9b8): I removed old JS code from Developer Portal default templates. No need to update the old ones, though. reCaptcha v3 should work fine for all existing templates.
- Update tests (https://github.com/3scale/porta/pull/3645/commits/db0a482a46889b037f6205c31a9d5589fa747350): Adapt existing tests to the new bot protection workflow
- Add tests for the new logic (https://github.com/3scale/porta/pull/3645/commits/2497f197fd3e524cf0eb051f06a40d8a0cf0089c): Add unit and integration tests for the new modules.
